typescript/sdk/src/hook/EvmHookModule.ts

```diff
@@ -1,53 +1,599 @@
-import { HookFactories } from './contracts.js';
-import { HookConfig } from './types.js';
+import { HyperlaneHookDeployer } from './HyperlaneHookDeployer.js';
+import { DeployedHook } from './contracts.js';
+import {
+  AggregationHookConfig,
+  DomainRoutingHookConfig,
+  FallbackRoutingHookConfig,
+  HookConfig,
+  HookType,
+  IgpHookConfig,
+  OpStackHookConfig,
+  PausableHookConfig,
+  ProtocolFeeHookConfig,
+} from './types.js';
+
+type HookModuleAddresses = {
+  deployedHook: Address;
+  mailbox: Address;
+  proxyAdmin: Address;
+};
 
-// WIP example implementation of EvmHookModule
 export class EvmHookModule extends HyperlaneModule<
   ProtocolType.Ethereum,
   HookConfig,
-  HyperlaneAddresses<HookFactories> & {
-    deployedHook: Address;
-  }
+  HyperlaneAddresses<ProxyFactoryFactories> & HookModuleAddresses
 > {
-  protected logger = rootLogger.child({ module: 'EvmHookModule' });
-  protected reader: EvmHookReader;
+  protected readonly logger = rootLogger.child({ module: 'EvmHookModule' });
+  protected readonly reader: EvmHookReader;
+
+  // Adding these to reduce how often we need to grab from MultiProvider.
+  public readonly chainName: string;
+  // We use domainId here because MultiProvider.getDomainId() will always
+  // return a number, and EVM the domainId and chainId are the same.
+  public readonly domainId: number;
 
   protected constructor(
     protected readonly multiProvider: MultiProvider,
-    protected readonly deployer: HyperlaneDeployer<any, any>,
+    protected readonly deployer: HyperlaneHookDeployer,
     args: HyperlaneModuleArgs<
       HookConfig,
-      HyperlaneAddresses<HookFactories> & {
-        deployedHook: Address;
-      }
+      HyperlaneAddresses<ProxyFactoryFactories> & HookModuleAddresses
     >,
   ) {
     super(args);
-    this.reader = new EvmHookReader(multiProvider, args.chain);
+    this.reader = new EvmHookReader(multiProvider, this.args.chain);
+
+    this.chainName = this.multiProvider.getChainName(this.args.chain);
+    this.domainId = this.multiProvider.getDomainId(this.chainName);
   }
 
   public async read(): Promise<HookConfig> {
-    return await this.reader.deriveHookConfig(this.args.addresses.deployedHook);
+    const config = await this.reader.deriveHookConfig(
+      this.args.addresses.deployedHook,
+    );
+    return normalizeConfig(config);
   }
 
-  public async update(_config: HookConfig): Promise<EthersV5Transaction[]> {
+  public async update(
+    _config: HookConfig,
+  ): Promise<AnnotatedEthersV5Transaction[]> {
     throw new Error('Method not implemented.');
   }
 
   // manually write static create function
-  public static create(_config: HookConfig): Promise<EvmHookModule> {
-    throw new Error('not implemented');
+  public static async create({
+    chain,
+    config,
+    deployer,
+    factories,
+    mailbox,
+    proxyAdmin,
+    multiProvider,
+  }: {
+    chain: ChainNameOrId;
+    config: HookConfig;
+    deployer: HyperlaneHookDeployer;
```

Comment:
why does this need to be passed in?

Reply:
it seems like a lot of code is redundant between this create fn and the hook deployer
when will we dedupe?

Reply:
only using fns from the abstract class, providing `HyperlaneHookDeployer` is a quick way to have a deployer type-compliant with the relevant hook/igp contracts

Reply:
naturally there will be a lot of dupe to avoid reinventing the wheel, part of the reason the updates are so difficult is because it involves understanding what's going on in the deployer and then doing a split